packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+import * as cxapi from '@aws-cdk/cx-api';
+import { CloudFormation, S3 } from 'aws-sdk';
+import * as chalk from 'chalk';
+import { ISDK, Mode, SdkProvider } from './aws-auth';
+import { print } from '../logging';
+import { DEFAULT_TOOLKIT_STACK_NAME, ToolkitInfo } from './toolkit-info';
+
+// Must use a require() otherwise esbuild complains
+// eslint-disable-next-line @typescript-eslint/no-require-imports
+const pLimit: typeof import('p-limit') = require('p-limit');
+
+const ISOLATED_TAG = 'aws-cdk:isolated';
+const P_LIMIT = 50;
+const DAY = 24 * 60 * 60 * 1000; // Number of milliseconds in a day
+
+class ActiveAssetCache {
+  private readonly stacks: Set<string> = new Set();
+
+  public rememberStack(stackTemplate: string) {
+    this.stacks.add(stackTemplate);
+  }
+
+  public contains(asset: string): boolean {
+    for (const stack of this.stacks) {
+      if (stack.includes(asset)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+class S3Asset {
+  private cached_tags: S3.TagSet | undefined = undefined;
+
+  public constructor(private readonly bucket: string, public readonly key: string, public readonly size: number) {}
+
+  public fileName(): string {
+    return this.key.split('.')[0];
+  }
+
+  public async allTags(s3: S3) {
+    if (this.cached_tags) {
+      return this.cached_tags;
+    }
+
+    const response = await s3.getObjectTagging({ Bucket: this.bucket, Key: this.key }).promise();
+    this.cached_tags = response.TagSet;
+    return response.TagSet;
+  }
+
+  private async getTag(s3: S3, tag: string) {
+    const tags = await this.allTags(s3);
+    return tags.find(t => t.Key === tag)?.Value;
+  }
+
+  private async hasTag(s3: S3, tag: string) {
+    const tags = await this.allTags(s3);
+    return tags.some(t => t.Key === tag);
+  }
+
+  public async noIsolatedTag(s3: S3) {
+    return !(await this.hasTag(s3, ISOLATED_TAG));
+  }
+
+  public async isolatedTagBefore(s3: S3, date: Date) {
+    const tagValue = await this.getTag(s3, ISOLATED_TAG);
+    print(chalk.red('a', tagValue, 'b', date));
+    if (!tagValue || tagValue == '') {
+      print(chalk.red('no tag'));
+      return false;
+    }
+    print(chalk.red('tag', typeof(tagValue)));
+    return new Date(tagValue) < date;
+  }
+}
+
+/**
+ * Props for the Garbage Collector
+ */
+interface GarbageCollectorProps {
+  /**
+   * The action to perform. Specify this if you want to perform a truncated set
+   * of actions available.
+   *
+   * @default 'full'
+   */
+  readonly action: 'print' | 'tag' | 'delete-tagged' | 'full';
```

Comment:
Also can we make this an enum the same way?
```suggestion
  readonly enum ActionType {
  PRINT = 'print',
  DELETE_TAGGED = 'delete-tagged',
  FULL = 'full'
}
```